Staking and other updates (#16)
subspace/subnomicon: docs/decex/staking.md

Any SSC token holder who has more than the minimum nominator stake (currently 1 SSC) may choose to join this operator’s pool by submitting the nomination extrinsic with the deposit amount of SSC they wish to stake. 

1. The amount of deposited SSC is added to the list of pending deposits within the operator's pool. 
2. At the end of an epoch (currently 100 blocks, ~ 10 minutes), the nominator's deposit is processed.

> 3. A part of the deposit is taken as a reserve towards a storage fee fund. This reserve is calculated as a percentage of the deposit (currently, 20%), and is used to pay for the storage fees of bundles created by the operator of the pool and does not affect the stake distribution. The reserved amount is transferred to the operators storage fee fund, while the rest of the deposit remains locked in the nominator's account. This amount is partially refunded with each withdrawal. 

Reviewer: ```suggestion
3. A part of the deposit is taken as a reserve towards a storage fee fund. This reserve is calculated as a percentage of the deposit (currently, 20%), and is used to pay for the storage fees of bundles created by the operator of the pool and does not affect the stake distribution. The reserved amount is transferred to the operator's storage fee fund, while the rest of the deposit remains locked in the nominator's account. This amount is partially refunded with each withdrawal. 
```

(assuming the storage fee fund is not shared by all operators?)

Reviewer: If I try to misunderstand these updates it's not clear why we need the storage fee reserve. My question would be "why are the storage fees not being charged to the user who submits the transaction?"

Author: You're right, I'll make this more clear. The storage fees are charged to the user, but only after the tx is executed on domain, however consensus chain requires front payment for storage and that's where the pool's storage fund comes in. Each operator pool has a separate fund.

When the nominator decides to withdraw their stake or fees, they submit a withdraw extrinsic. The withdraw extrinsic is processed at the end of the epoch and the stake is removed from the operator's pool and the domain's total stake. The nominator is then entitled to the fees percentage based on the stake shares and the amount of time they have staked.

Operators can also withdraw their stake and fees at any time by submitting a `withdraw_stake` extrinsic. Operators who wish to withdraw all of their stake and earned fees have to submit a deregistration extrinsic, as it is forbidden to withdraw below the domain's minimum stake requirements. The deregistered operator will be removed from the domain and their stake and the stakes of all nominators will be returned to their accounts.

<div align="center">
    <img src="/img/Nomination_Pool-light.svg#gh-light-mode-only" alt="Nomination_Pool" />
    <img src="/img/Nomination_Pool-dark.svg#gh-dark-mode-only" alt="Nomination_Pool" />
</div>

### Unlocking Funds

> Withdrawals have a lock period of 2 days (currently 28 800 blocks, ~ 48 hours). After the locking period, the withdrawn amount can be unlocked in the users account with the `unlock_funds` extrinsic. All withdrawals requested in the same stake epoch are aggregated together and the total amount is unlocked at once. This locking period is necessary to ensure that the domain block executing the withdrawal is confirmed and not challenged by a fraud proof and to increase the economic stability of domains. 

Reviewer: ```suggestion
Withdrawals have a lock period of 2 days (currently 28 800 blocks, ~ 48 hours). After the locking period, the withdrawn amount can be unlocked in the user's account with the `unlock_funds` extrinsic. All withdrawals requested in the same stake epoch are aggregated together and the total amount is unlocked at once. This locking period is necessary to ensure that the domain block executing the withdrawal is confirmed and not challenged by a fraud proof in order to increase the economic stability of domains. 
```

> Withdrawals have a lock period of 2 days (currently 28 800 blocks, ~ 48 hours). After the locking period, the withdrawn amount can be unlocked in the user's account with the `unlock_funds` extrinsic. All withdrawals requested in the same stake epoch are aggregated together and the total amount is unlocked at once. This locking period is necessary to ensure that the domain block executing the withdrawal is confirmed and not challenged by a fraud proof in order to increase the economic stability of domains. 

Reviewer 2: ```suggestion
Withdrawals have a lock period of 1 day (currently 14 400 blocks, ~ 24 hours). After the locking period, the withdrawn amount can be unlocked in the user's account with the `unlock_funds` extrinsic. All withdrawals requested in the same stake epoch are aggregated together and the total amount is unlocked at once. This locking period is necessary to ensure that the domain block executing the withdrawal is confirmed and not challenged by a fraud proof in order to increase the economic stability of domains. 
```

Reviewer: Shouldn't this say 1 day?

Withdrawals have a lock period of 2 days -> Withdrawals have a lock period of 1 day

Reviewer 2: Right, edited the initial message!

Author: No, it's actually 2 days until the user can unlock and use the funds. The reason is that `BlockTreePruningDepth` and `StakeWithdrawalLockingPeriod` are added ([see code here](https://github.com/subspace/subspace/blob/deeec211e11b75d1280e95ceb500722c9e892291/crates/pallet-domains/src/staking.rs#L690). Basically `StakeWithdrawalLockingPeriod` kicks in after the block with withdrawal request has been confirmed (=`BlockTreePruningDepth`has passed). So it's 2 days in total. And will probably be 7 by mainnet.

Reviewer 2: I see, thanks for the clarification!

### Example

Operator $O$ has staked 100 SSC and registered as an operator with minimum nominator stake of 10 SSC and nomination tax of 5%. The required storage fee reserve deposit is 20%. Operator $O$ has 2 nominators $N_1$ and $N_2$ each staked 50 SSC and reserved 20% = 10 SSC each for storage fees. Initially $\text{shares\_per\_ssc} = 1$, so $O$ gets 80 shares, and $N_1$ and $N_2$ each get 40 shares and $\text{total\_shares}=80+40+40=160$ in the stake and the same shares in the storage fee reserve. 

In the next epoch, the pool has earned 20 SSC of compute fees and refunded an extra 4 SSC of storage fees, and the operator took 5% of compute fees as a commission (1 SSC). The pool stake is now 160+20=180 SSC and storage reserve is now 40+4=44 SSC.

> The pool end-of-epoch $\text{shares\_per\_ssc}$ is now $160/(160 + 20 * (1-0.05)) = 0.893855$. Notice that 4 SSC of storage fees do not count into $\text{shares\_per\_ssc}$ calculation, which allows us to sustain stable stake distribution despite fluctuating size of storage fee reserve. 

Reviewer: ```suggestion
The pool end-of-epoch $\text{shares\_per\_ssc}$ is now $160/(160 + 20 * (1-0.05)) = 0.893855$. Notice that 4 SSC of storage fees do not count into $\text{shares\_per\_ssc}$ calculation, which allows us to sustain stable stake distribution despite the fluctuating size of the storage fee reserve. 
```

If a new nominator $N_3$ stakes 33.6 SSC, 6.72 SSC will be reserved for storage fee fund, and the $\text{shares}$ they will get is $((33.6-6.72) * 0.893855) = 24$. The pool total stake becomes $180+26.88=206.88$ SSC, total shares $160+24+1=185$ and storage fee reserve 50.72 SSC.

> Suppose after some time $\text{shares\_per\_ssc}$ value of this pool becomes 0.8 and the storage fee fund balance is 52 SSC. Suppose $N_1$ wants to "sell" $\text{withdraw\_shares}=20$ shares. At the end of epoch, the 20 shares will be unstaked, and the corresponding amount of $20/0.8=25$ SSC will be deducted from the pool's total stake. The total amount of credits $N_1$ will get is $\text{withdraw\_shares}/\text{shares\_per\_ssc}+\text{storage\_fund\_balance}*\text{withdraw\_shares}/\text{total\_shares}=25 + 52*20/185=30.62$ SSC.

Reviewer: ```suggestion
Suppose after some time $\text{shares\_per\_ssc}$ value of this pool becomes 0.8 and the storage fee fund balance is 52 SSC. Suppose $N_1$ wants to "sell" $\text{withdraw\_shares}=20$ shares. At the end of the epoch, the 20 shares will be unstaked, and the corresponding amount of $20/0.8=25$ SSC will be deducted from the pool's total stake. The total amount of credits $N_1$ will get is $\text{withdraw\_shares}/\text{shares\_per\_ssc}+\text{storage\_fund\_balance}*\text{withdraw\_shares}/\text{total\_shares}=25 + 52*20/185=30.62$ SSC.
```

When the nominator decides to withdraw their stake or fees, they submit a withdraw extrinsic. The withdraw extrinsic is processed at the end of the epoch and the stake is removed from the operator's pool and the domain's total stake. The nominator is then entitled to the fees percentage based on the stake shares and the amount of time they have staked.

Operators can also withdraw their stake and fees at any time by submitting a `withdraw_stake` extrinsic. Operators who wish to withdraw all of their stake and earned fees have to submit a deregistration extrinsic, as it is forbidden to withdraw below the domain's minimum stake requirements. The deregistered operator will be removed from the domain and their stake and the stakes of all nominators will be returned to their accounts.

<div align="center">
    <img src="/img/Nomination_Pool-light.svg#gh-light-mode-only" alt="Nomination_Pool" />
    <img src="/img/Nomination_Pool-dark.svg#gh-dark-mode-only" alt="Nomination_Pool" />
</div>

### Unlocking Funds

> Withdrawals have a lock period of 2 days (currently 28 800 blocks, ~ 48 hours). After the locking period, the withdrawn amount can be unlocked in the user's account with the `unlock_funds` extrinsic. All withdrawals requested in the same stake epoch are aggregated together and the total amount is unlocked at once. This locking period is necessary to ensure that the domain block executing the withdrawal is confirmed and not challenged by a fraud proof in order to increase the economic stability of domains. 

Reviewer 2: I would like to clarify my understanding here. 
Previously, in order to unlock tokens operators:
1 -  Can either withdraw tokens using `withdrawStake(operatorId, withdraw)`
2 - Or can withdraw stake by de-registering `deregisterOperator(operatorID)`
In both cases their funds were automatically unlocked after the locking period. 
Would that change at all now (other than the locking time changes)? It sounds like the funds won't be automatically unlocked, but users have to use another extinsics call to `unlock_funds ` for the tokens to be released. Maybe I'm misinterpreting something.

Author: Yes. After 1 or 2, the user has to wait and then submit `unlock_funds` to actually be able to use the tokens. `withdrawStake` now basically only "unstakes" from the pool, but doesn't actually compute the individual amount or mint/transfer. This change is worse UX but a huge optimization to domain that lets us avoid iterating through all hundreds (or thousands) of nominators every time.
Currently, the `unlock_funds` has to be submitted manually, but we could have it automated by the staking interface or wallet so that the funds that are ready to be unlocked show like a reminder icon for the user or even submitted automatically as `future` transaction. There's room for front-end UX improvement for sure.

Reviewer 2: Makes sense, thank you!
I'm adding the step to the guide as well, at least until we automate it.

## Power Balance 

> Farmers who have earned storage rewards nominate operators to execute transactions. This nomination system balances the power between farmers and operators, and both parties share the fees and the potential penalties (slashing). Farmer-nominated operators get a higher chance to produce blocks proportional to the amount of stake backing them, thus, higher revenues. Farmers have the power to nominate operators they trust to execute transactions properly. On the other hand, operators compete to be nominated by providing good service, maintaining a good reputation within the community, and having reasonable commission. 

Reviewer: > balances the power between farmers and operators, and both parties share the fees and the potential penalties (slashing)

This reads as though farmers can be slashed. I assume you mean something like "token holders who nominate"?

> Farmer-nominated operators

And maybe we should switch from implying only farmers can nominate here? We should make it clear it's a use-case for acquiring/holding the token?

I get you're explaining the balance of power but, in isolation, this passage reads as though non-farmer token-holders do not exist.

Reviewer: I re-read the notes above about who can be a nominator. Maybe my concerns here are unfounded - your shout :)

Author: I'll try to revise this section. Maybe repeat the sentence that anyone can nominate.
